[IsDeprecatedWeakRefSmartPointerException] Make WebRemoteObjectRegistry ref-counted

RupinMittal opened the pull request:
#### b15f86cdf39f27860179698c6d67898c43dce4f4
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make WebRemoteObjectRegistry ref-counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=281263">https://bugs.webkit.org/show_bug.cgi?id=281263</a>
<a href="https://rdar.apple.com/137716730">rdar://137716730</a>

Reviewed by Chris Dumez.

Remove IsDeprecatedWeakRefSmartPointerException] by making
WebRemoteObjectRegistry ref-counted.

* Source/WebKit/Shared/API/Cocoa/RemoteObjectRegistry.h:
* Source/WebKit/Shared/API/Cocoa/RemoteObjectRegistry.mm:
(WebKit::RemoteObjectRegistry::send):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
(-[_WKRemoteObjectRegistry _initWithWebPage:]):
(-[_WKRemoteObjectRegistry _initWithWebPageProxy:]):
(-[_WKRemoteObjectRegistry _sendInvocation:interface:]):
(-[_WKRemoteObjectRegistry _invokeMethod:]):
* Source/WebKit/UIProcess/Cocoa/UIRemoteObjectRegistry.cpp:
(WebKit::UIRemoteObjectRegistry::backgroundActivity):
(WebKit::UIRemoteObjectRegistry::sendInvocation):
(WebKit::UIRemoteObjectRegistry::messageSender):
(WebKit::UIRemoteObjectRegistry::messageDestinationID):
(WebKit::UIRemoteObjectRegistry::protectedPage): Deleted.
* Source/WebKit/UIProcess/Cocoa/UIRemoteObjectRegistry.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.cpp:
(WebKit::WebRemoteObjectRegistry::close):
(WebKit::WebRemoteObjectRegistry::messageSender):
(WebKit::WebRemoteObjectRegistry::messageDestinationID):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.h:

Canonical link: <a href="https://commits.webkit.org/285150@main">https://commits.webkit.org/285150@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95e3e03111e1c2d622a322a50fc1c345e30e0baf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71701 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51114 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24475 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22906 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58915 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22726 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/75816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/15110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46373 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61767 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/75816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43034 "Passed tests") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/21247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64939 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19597 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77535 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15935 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15979 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61800 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64342 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15846 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12503 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46914 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47985 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49269 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47727 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
```diff
@@ -175,10 +175,11 @@ - (void)_sendInvocation:(NSInvocation *)invocation interface:(_WKRemoteObjectInt
     [encoder encodeObject:invocation forKey:invocationKey];
 
-    if (!_remoteObjectRegistry)
+    RefPtr protectedRemoteObjectRegistry = _remoteObjectRegistry;
```
I think we can just name this `remoteObjectRegistry`. We avoid `protected` in local variable names as much as possible nowadays.

cdumez commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
```diff
@@ -270,10 +271,11 @@ - (void)_invokeMethod:(const WebKit::RemoteObjectInvocation&)remoteObjectInvocat
                 // keep track of all reply block checkers and invalidate them (sending the unused reply message) in
                 // -[_WKRemoteObjectRegistry _invalidate].
-                if (!m_remoteObjectRegistry->_remoteObjectRegistry)
+                RefPtr protectedRemoteObjectRegistry = m_remoteObjectRegistry->_remoteObjectRegistry;
```
ditto.

cdumez commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
```diff
@@ -295,8 +297,9 @@ - (void)_invokeMethod:(const WebKit::RemoteObjectInvocation&)remoteObjectInvocat
 
-            if (remoteObjectRegistry->_remoteObjectRegistry)
-                remoteObjectRegistry->_remoteObjectRegistry->sendReplyBlock(replyID, WebKit::UserData([encoder rootObjectDictionary]));
+            RefPtr protectedRemoteObjectRegistry = remoteObjectRegistry->_remoteObjectRegistry;
```
ditto.

RupinMittal replied:
I changed  `protectedRemoteObjectRegistry` to `remoteObjectRegistry` in the other locations but had to keep it here since `remoteObjectRegistry` is already being used in this scope.

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.h:
```diff
@@ -26,31 +26,32 @@
+
     MessageSender messageSender() final;
     uint64_t messageDestinationID() final;
 
```
Well, same old bug. The WeakRef below needs to become a WeakPtr with some null check since you're extending lifetime.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/UIRemoteObjectRegistry.h:
```diff
@@ -26,22 +26,32 @@
+
     Ref<WebPageProxy> protectedPage();
     MessageSender messageSender() final;
     uint64_t messageDestinationID() final;
```
Well, same old bug. The WeakRef below needs to become a WeakPtr with some null check since you're extending lifetime.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
```diff
@@ -295,8 +297,9 @@ - (void)_invokeMethod:(const WebKit::RemoteObjectInvocation&)remoteObjectInvocat
-            if (remoteObjectRegistry->_remoteObjectRegistry)
-                remoteObjectRegistry->_remoteObjectRegistry->sendReplyBlock(replyID, WebKit::UserData([encoder rootObjectDictionary]));
+            RefPtr protectedRemoteObjectRegistry = remoteObjectRegistry->_remoteObjectRegistry;
+            if (protectedRemoteObjectRegistry)
```
`if (RefPtr protectedRemoteObjectRegistry = remoteObjectRegistry->_remoteObjectRegistry)`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.cpp:
```diff
@@ -50,21 +50,28 @@ WebRemoteObjectRegistry::~WebRemoteObjectRegistry()
+auto WebRemoteObjectRegistry::messageSender() -> std::optional<MessageSender>
 {
-    return m_page.get();
+    if (RefPtr page = m_page.get())
```
I don't think we need a RefPtr here.

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.cpp:
```diff
@@ -50,21 +50,28 @@ WebRemoteObjectRegistry::~WebRemoteObjectRegistry()
+std::optional<uint64_t> WebRemoteObjectRegistry::messageDestinationID()
 {
-    return m_page.get().webPageProxyIdentifier().toUInt64();
+    if (RefPtr page = m_page.get())
```
I doubt we need a RefPtr here. `webPageProxyIdentifier()` seems trivial.

cdumez reviewed (approved):
LGTM with nits.

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Make WebRemoteObjectRegistry ref-counted
https://bugs.webkit.org/show_bug.cgi?id=281263
rdar://137716730

Reviewed by Chris Dumez.

Remove IsDeprecatedWeakRefSmartPointerException] by making
WebRemoteObjectRegistry ref-counted.

* Source/WebKit/Shared/API/Cocoa/RemoteObjectRegistry.h:
* Source/WebKit/Shared/API/Cocoa/RemoteObjectRegistry.mm:
(WebKit::RemoteObjectRegistry::send):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectRegistry.mm:
(-[_WKRemoteObjectRegistry _initWithWebPage:]):
(-[_WKRemoteObjectRegistry _initWithWebPageProxy:]):
(-[_WKRemoteObjectRegistry _sendInvocation:interface:]):
(-[_WKRemoteObjectRegistry _invokeMethod:]):
* Source/WebKit/UIProcess/Cocoa/UIRemoteObjectRegistry.cpp:
(WebKit::UIRemoteObjectRegistry::backgroundActivity):
(WebKit::UIRemoteObjectRegistry::sendInvocation):
(WebKit::UIRemoteObjectRegistry::messageSender):
(WebKit::UIRemoteObjectRegistry::messageDestinationID):
(WebKit::UIRemoteObjectRegistry::protectedPage): Deleted.
* Source/WebKit/UIProcess/Cocoa/UIRemoteObjectRegistry.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.cpp:
(WebKit::WebRemoteObjectRegistry::close):
(WebKit::WebRemoteObjectRegistry::messageSender):
(WebKit::WebRemoteObjectRegistry::messageDestinationID):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebRemoteObjectRegistry.h:

Canonical link: https://commits.webkit.org/285150@main
